### PR TITLE
Custom CA/certs, HSTS test, SHA-1 2017 site, 480/512 bit DH sites

### PR DESCRIPTION
**Pull request by marumari**
Created a root CA, intermediate CA, and all of the badssl.com certs that  are used currently.  Also created a SHA-1 (2017) cert, which isn't currently tested, as well as a certificate generator for everything together.

### CODE REVIEW

**Comment by lgarron** [comment]
The script is fantastic. :-D :+1:

However, I need to be able to deploy to the web using the real certs. Could you configure it so that `nginx` uses the "real" certs if [server condition] holds, else to use the self-signed certs?

A simple condition that holds on the current live server: `hostname` (e.g. run from the commandline) is `badssl-com`. Another option would be to symlink cert folders based on the deployment.
I'd merge this and think of something myself, but I don't quite have the time, and I don't accidentally want to serve a wrong cert on `sha1-2017.badssl.com`. Could you try to think of something?

File: certs/cert-generator/cert-generator.sh [hunk 1]
```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env bash
+
```
Comment (lgarron) [inline certs/cert-generator/cert-generator.sh]: I like my bash scripts to be defensive. Would you mind adding the following?

    set -eu
    cd "$(dirname ${0})"

**Comment by marumari** [comment]
Right now, the script just creates the certs in the self-signed directory.  It's up to the developer to actually copy them into the certs directory if they want to use them; this way, I wouldn't tread all over your own certs.  As for sha-2017, as long as you don't link it on the main site until you have a true cert, you should probably be fine.  :)

Anyways, I could add a bit to the Makefile for the self-signed certs?  I don't think I can run based on the hostname: I currently test with badssl.com pointed at my own local VM.

File: certs/cert-generator/cert-generator.sh (hunk 1 above)
Comment (marumari) [inline certs/cert-generator/cert-generator.sh]: Done!  I had been meaning to do that, just skipped my mind amidst all the glowering at the humdrumness that is openssl.  :)

**Comment by lgarron** [comment]
> As for sha-2017, as long as you don't link it on the main site until you have a true cert, you should probably be fine. :)

Unfortunately, nginx breaks because `sha1-2017.badssl.com.conf` references a non-existent key. :-(

> Anyways, I could add a bit to the Makefile for the self-signed certs? I don't think I can run based on the hostname: I currently test with badssl.com pointed at my own local VM.

By `hostname`, I actually mean the name that the server gives you if you run `hostname` on the commandline. My Ubuntu VM gives me `ubuntu`.

**Comment by marumari** [comment]
Oh, I see what you're saying.  That's a fine idea!  I'll add that it to the pile.  :)

As for sha1-2017, can't you just copy wildcard.sha1-2016.pem to wildcard.sha1-2017.pem as a placeholder, until you get a true SHA-1 cert that expires in 2017?  That should help with the nginx problems.

**Comment by marumari** [comment]
Okay, I just pushed the change to automatically deploy the certs into the certs directory if it isn't badssl-com.  :)

**Comment by lgarron** [comment]
Sorry for the delay on this; I want to put a temporary shim on the server to avoid serving sha1-2017 from it, but I haven't had time.

**Comment by marumari** [comment]
Okay, just added support for broken/weak DH.

Broken (480): Firefox and Chrome error out (ssl_error_weak_server_ephemeral_dh_key), Safari gives a broken lock icon, but continues otherwise, IE doesn't seem to care.

Weak (512): Firefox, Chrome, and IE continue as normal, Safari gives a broken lock icon but continues otherwise.

WTF, Mozilla / Google / Microsoft?

**Comment by lgarron** [comment]
I did the following:
- Merged the custom CA and `sha1-2017` commits from this PR.
- Added an HSTS test in 452287623f4d466e80a1e722eda3b51e1485531c.
- Cherry-picked the DH commit (as 44df428186b432583cee701df388184726f54a4b) and changed it a bit in aea4d8e05fadd8edc42db127e9cae1ffdaec9924.

I handled the fact that we don't have a real `sha1-2017` cert by [simply not deploying that domain](25b5642c8bb03a15765922d1fe82840b9c2faf9e).

I *believe* that covers everything in this commit.
Thanks for a lot of good code. :-)

**Pull request closed without merging**